File: everything/src/tools/create-ghost-post.js
/**
 * Tool handler for creating a new post in Ghost blog
 */
import axios from 'axios';
import crypto from 'crypto';

export async function handleCreateGhostPost(server, args) {
    try {
        // Special case for tool info request (handled by tool definition)
        
        // Validate required parameters
        if (!args.title || !args.content) {
            return {
                content: [{
                    type: 'text',
                    text: JSON.stringify({ error: "Title and content are required" }, null, 2),
                }],
                isError: true
            };
        }

        // Set default values
        const status = args.status || "draft";
        const tags = args.tags || null;
        const featured = args.featured !== undefined ? args.featured : false;

        // Get API credentials
        const apiUrl = process.env.GHOST_API_URL || "https://blog.emmanuelu.com";
        const adminId = "67b2d2824fdabf0001eb99ea";
        const adminSecret = "100eda163e9fea6906fbd7ddc841812c70561b19ab2e6d7b31f844f6ccd7b05d";

        try {
            // Create JWT token manually
            const header = {
                alg: "HS256",
                typ: "JWT",
                kid: adminId
            };
            
            const iat = Math.floor(Date.now() / 1000);
            const payload = {
                iat: iat,
                exp: iat + 300,  // Token expires in 5 minutes
                aud: "/admin/"
            };

            // Base64 encode header and payload
            const headerEncoded = Buffer.from(JSON.stringify(header))
                .toString('base64')
                .replace(/=+$/, '')
                .replace(/\+/g, '-')
                .replace(/\//g, '_');
                
            const payloadEncoded = Buffer.from(JSON.stringify(payload))
                .toString('base64')
                .replace(/=+$/, '')
                .replace(/\+/g, '-')
                .replace(/\//g, '_');
            
            // Create signature
            const message = `${headerEncoded}.${payloadEncoded}`;
            const key = Buffer.from(adminSecret, 'hex');
            const signature = crypto.createHmac('sha256', key)
                .update(message)
                .digest('base64')
                .replace(/=+$/, '')
                .replace(/\+/g, '-')
                .replace(/\//g, '_');
            
            // Combine into final token
            const token = `${headerEncoded}.${payloadEncoded}.${signature}`;

            // Set up axios instance with retry logic
            const axiosInstance = axios.create({
                timeout: 30000,
                headers: {
                    'Authorization': `Ghost ${token}`,
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                }
            });

            // Add retry logic
            axiosInstance.interceptors.response.use(null, async (error) => {
                const { config } = error;
                if (!config || !config.retry) {
                    return Promise.reject(error);
                }
                
                config.retryCount = config.retryCount || 0;
                
                if (config.retryCount >= config.retry) {
                    return Promise.reject(error);
                }
                
                config.retryCount += 1;
                
                const backoff = config.retryDelay || 1000;
                await new Promise(resolve => setTimeout(resolve, backoff));
                
                return axiosInstance(config);
            });

            // Ghost Admin API endpoint
            const url = `${apiUrl.replace(/\/$/, '')}/ghost/api/admin/posts/?source=html`;

            // Prepare request payload
            const requestPayload = {
                posts: [{
                    title: args.title,
                    html: args.content,
                    status: status,
                    featured: featured
                }]
            };

            if (tags) {
                requestPayload.posts[0].tags = tags;
            }

            console.log(`Making request to: ${url}`);
            console.log(`Headers: ${JSON.stringify(axiosInstance.defaults.headers)}`);
            console.log(`Payload: ${JSON.stringify(requestPayload, null, 2)}`);

            // Make the API request
            const response = await axiosInstance.post(url, requestPayload, {
                retry: 3,
                retryDelay: 1000
            });

            return {
                content: [{
                    type: 'text',
                    text: JSON.stringify(response.data, null, 2),
                }],
            };
        } catch (error) {
            let errorMsg = error.message;
            
            if (error.response) {
                errorMsg += `\nResponse: ${JSON.stringify(error.response.data)}`;
            }
            
            return {
                content: [{
                    type: 'text',
                    text: JSON.stringify({ error: `Network or HTTP error - ${errorMsg}` }, null, 2),
                }],
                isError: true
            };
        }
    } catch (error) {
        return server.createErrorResponse(error);
    }
}

/**
 * Tool definition for create_ghost_post
 */
export const createGhostPostToolDefinition = {
    name: 'create_ghost_post',
    description: 'Creates a new post in Ghost blog',
    inputSchema: {
        type: 'object',
        properties: {
            title: {
                type: 'string',
                description: 'The title of the post',
            },
            content: {
                type: 'string',
                description: 'The content/body of the post (HTML)',
            },
            status: {
                type: 'string',
                description: 'Post status (draft, published, scheduled)',
                default: 'draft'
            },
            tags: {
                type: 'array',
                items: {
                    type: 'object'
                },
                description: 'List of tag objects (each with \'name\')',
            },
            featured: {
                type: 'boolean',
                description: 'Whether this is a featured post',
                default: false
            },
        },
        required: ['title', 'content'],
    },
};

File: src/tools/create_ghost_post.py
import json
import requests
import os
import base64
import hashlib
import hmac
import time
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry


def create_ghost_post(title: str, content: str, status: str = "draft", tags: list = None, featured: bool = False) -> str:
    """
    Creates a new post in Ghost blog.

    This function supports two usage scenarios:
      1) Creating a post
      2) Retrieving the tool's metadata by passing title="__tool_info__"

    Args:
        title (str): The title of the post
        content (str): The content/body of the post (HTML)
        status (str, optional): Post status (draft, published, scheduled)
        tags (list, optional): List of tag objects (each with 'name')
        featured (bool, optional): Whether this is a featured post

    Returns:
        str: JSON-formatted string containing the response
    """
    if title == "__tool_info__":
        info = {
            "name": "create_ghost_post",
            "description": "Creates a new post in Ghost blog",
            "args": {
                "title": {
                    "type": "string",
                    "description": "The title of the post",
                    "required": True
                },
                "content": {
                    "type": "string",
                    "description": "The content/body of the post (HTML)",
                    "required": True
                },
                "status": {
                    "type": "string",
                    "description": "Post status (draft, published, scheduled)",
                    "required": False,
                    "default": "draft"
                },
                "tags": {
                    "type": "list",
                    "description": "List of tag objects (each with 'name')",
                    "required": False
                },
                "featured": {
                    "type": "boolean",
                    "description": "Whether this is a featured post",
                    "required": False,
                    "default": False
                }
            }
        }
        return json.dumps(info)

    if not title or not content:
        return json.dumps({"error": "Title and content are required"})

    api_url = os.environ.get("GHOST_API_URL", "https://blog.emmanuelu.com").rstrip("/")
    admin_id = "67b2d2824fdabf0001eb99ea"
    admin_secret = "100eda163e9fea6906fbd7ddc841812c70561b19ab2e6d7b31f844f6ccd7b05d"

    try:
        # Create JWT token manually
        header = {
            "alg": "HS256",
            "typ": "JWT",
            "kid": admin_id
        }
        iat = int(time.time())
        payload = {
            "iat": iat,
            "exp": iat + 300,  # Token expires in 5 minutes
            "aud": "/admin/"
        }

        # Base64 encode header and payload
        header_encoded = base64.urlsafe_b64encode(json.dumps(header).encode()).rstrip(b'=').decode('utf-8')
        payload_encoded = base64.urlsafe_b64encode(json.dumps(payload).encode()).rstrip(b'=').decode('utf-8')
        
        # Create signature
        message = f"{header_encoded}.{payload_encoded}"
        key = bytes.fromhex(admin_secret)
        signature = hmac.new(key, message.encode(), hashlib.sha256).digest()
        signature_encoded = base64.urlsafe_b64encode(signature).rstrip(b'=').decode('utf-8')
        
        # Combine into final token
        token = f"{header_encoded}.{payload_encoded}.{signature_encoded}"

        session = requests.Session()
        adapter = HTTPAdapter(max_retries=Retry(
            total=3,
            backoff_factor=1,
            status_forcelist=[429, 500, 502, 503, 504]
        ))
        session.mount("http://", adapter)
        session.mount("https://", adapter)

        headers = {
            "Authorization": f"Ghost {token}",
            "Accept": "application/json",
            "Content-Type": "application/json"
        }

        # Ghost Admin API endpoint
        url = f"{api_url}/ghost/api/admin/posts/?source=html"

        payload = {
            "posts": [{
                "title": title,
                "html": content,
                "status": status,
                "featured": featured
            }]
        }

        if tags:
            payload["posts"][0]["tags"] = tags

        print(f"Making request to: {url}")
        print(f"Headers: {headers}")
        print(f"Payload: {json.dumps(payload, indent=2)}")

        response = session.post(
            url,
            headers=headers,
            json=payload,
            timeout=30
        )
        response.raise_for_status()
        return json.dumps(response.json())
    except requests.exceptions.RequestException as e:
        error_msg = str(e)
        if hasattr(e, 'response') and e.response is not None:
            error_msg += f"\nResponse: {e.response.text}"
        return json.dumps({"error": f"Network or HTTP error - {error_msg}"})
    except Exception as e:
        return json.dumps({"error": f"Unexpected error - {str(e)}"})
    finally:
        session.close()

if __name__ == "__main__":
    print("Testing create_ghost_post tool...")
    
    # Example post creation
    print("\nCreating test post...")
    try:
        new_post = create_ghost_post(
            title="Test Post",
            content="<p>This is a <strong>bold</strong> text, an <em>italic</em> text, and some <u>underlined</u> text.</p>",
            status="draft",
            tags=[{"name": "test"}],
            featured=False
        )
        response = json.loads(new_post)
        if "error" in response:
            print(f"Error creating post: {response['error']}")
        else:
            print("Post created successfully!")
            print(json.dumps(response, indent=2))
    except Exception as e:
        print(f"Error during test: {str(e)}")